crypto/s2n_hash.c

```diff
@@ -355,80 +322,95 @@ S2N_RESULT s2n_hash_state_validate(struct s2n_hash_state *state)
+    POSIX_ENSURE_REF(state->hash_impl);
+    POSIX_ENSURE_REF(state->hash_impl->init);
+    POSIX_GUARD(state->hash_impl->init(state, alg));
+
+    state->alg = alg;
+    state->is_ready_for_input = 1;
+    state->currently_in_hash = 0;
+    return S2N_SUCCESS;
 }
 
 int s2n_hash_update(struct s2n_hash_state *state, const void *data, uint32_t size)
 {
     POSIX_PRECONDITION(s2n_hash_state_validate(state));
     POSIX_ENSURE(S2N_MEM_IS_READABLE(data, size), S2N_ERR_PRECONDITION_VIOLATION);
+    POSIX_ENSURE(state->is_ready_for_input, S2N_ERR_HASH_NOT_READY);
+
+    POSIX_ENSURE_REF(state->hash_impl);
     POSIX_ENSURE_REF(state->hash_impl->update);
+    POSIX_GUARD(state->hash_impl->update(state, data, size));
 
-    return state->hash_impl->update(state, data, size);
+    POSIX_ENSURE(size <= (UINT64_MAX - state->currently_in_hash), S2N_ERR_INTEGER_OVERFLOW);
+    state->currently_in_hash += size;
+    return S2N_SUCCESS;
 }
 
 int s2n_hash_digest(struct s2n_hash_state *state, void *out, uint32_t size)
 {
     POSIX_PRECONDITION(s2n_hash_state_validate(state));
     POSIX_ENSURE(S2N_MEM_IS_READABLE(out, size), S2N_ERR_PRECONDITION_VIOLATION);
+    POSIX_ENSURE(state->is_ready_for_input, S2N_ERR_HASH_NOT_READY);
+
+    POSIX_ENSURE_REF(state->hash_impl);
     POSIX_ENSURE_REF(state->hash_impl->digest);
+    POSIX_GUARD(state->hash_impl->digest(state, out, size));
 
-    return state->hash_impl->digest(state, out, size);
+    state->currently_in_hash = 0;
+    state->is_ready_for_input = 0;
+    return S2N_SUCCESS;
 }
 
 int s2n_hash_copy(struct s2n_hash_state *to, struct s2n_hash_state *from)
 {
     POSIX_PRECONDITION(s2n_hash_state_validate(to));
     POSIX_PRECONDITION(s2n_hash_state_validate(from));
+
+    POSIX_ENSURE_REF(from->hash_impl);
     POSIX_ENSURE_REF(from->hash_impl->copy);
+    POSIX_GUARD(from->hash_impl->copy(to, from));
 
-    return from->hash_impl->copy(to, from);
+    to->hash_impl = from->hash_impl;
+    to->alg = from->alg;
+    to->is_ready_for_input = from->is_ready_for_input;
+    to->currently_in_hash = from->currently_in_hash;
+    return S2N_SUCCESS;
 }
 
 int s2n_hash_reset(struct s2n_hash_state *state)
 {
     POSIX_ENSURE_REF(state);
-    /* Ensure that hash_impl is set, as it may have been reset for s2n_hash_state on s2n_connection_wipe.
-     * When in FIPS mode, the EVP API's must be used for hashes.
-     */
-    POSIX_GUARD(s2n_hash_set_impl(state));
-
+    POSIX_ENSURE_REF(state->hash_impl);
     POSIX_ENSURE_REF(state->hash_impl->reset);
-
-    return state->hash_impl->reset(state);
+    POSIX_GUARD(state->hash_impl->reset(state));
+    POSIX_GUARD(s2n_hash_init(state, state->alg));
+    return S2N_SUCCESS;
 }
 
 int s2n_hash_free(struct s2n_hash_state *state)
 {
     if (state == NULL) {
         return S2N_SUCCESS;
     }
-    /* Ensure that hash_impl is set, as it may have been reset for s2n_hash_state on s2n_connection_wipe.
-     * When in FIPS mode, the EVP API's must be used for hashes.
-     */
```

Comment:
I'm convinced that this isn't currently true, and I'm not sure it was ever true? Wiping a connection doesn't erase the hash_impl-- and if it somehow did, it'd also likely erase all the other state that hash_impl->free is supposed to cleanup and therefore cause a memory leak. hash_impl should be set on _new, when the implementation specific memory is allocated, and then never touched again.

The previous two hash_impls could be chosen between when s2n_init was called, so calling s2n_hash_set_impl repeatedly wasn't an issue. But if we're going to add more complex logic around choosing different hash_impls, then we need to be more clear about the s2n_hash_state lifecycle.